### DISCUSSION TITLE
I see Wireguard, can Tailscale be supported?

### DISCUSSION THREAD

**Discussion by thereal-BLKMGK** [themactep/thingino-firmware#233] [Ideas]
Tailscale is based on Wireguard but makes setup even easier. Would adding Tailscale be possible? It's able to run on routers so hopefully the processor wouldn't be an issue.

**Comment by gtxaspec** [comment]
tailscale works, but the binaries are huge, exceeding the total capacity of the onboard flash storage chips.  you can download the binaries to an sd card and run it from there.

**Reply by thereal-BLKMGK** [reply]
Oh duh, you're right it's a pig and I've been used to putting it on things with much more storage. I like your idea of using the SD card though! I've ordered up a pile of cards and will hopefully be able to poke at it tomorrow - thank you for pointing that out! I'll report back as to how it goes, hopefully as smooth as my Unifi router went :) I'll be trying a microUSB ethernet\POE thingy next!

**Comment by numericOverflow** [comment]
@gtxaspec - Obviously it's still hefty, but looks like the tailscale docs claim to get the binary down to ~5mb

https://tailscale.com/kb/1207/small-tailscale

Do you know if their suggestions were used when you tested the binary in these commits, or did you just want to avoid the idea of a packer completely for all the reasons?

- add thingino-tailscale package https://github.com/themactep/thingino-firmware/commit/4eef7cf363ae4a47eef58bd392b47a0c6005c8e1
- tailscale: drop package due to its huge size https://github.com/themactep/thingino-firmware/commit/3fdeadb59595c979798528a81fa43f4431445c32

**Comment by gtxaspec** [comment]
i can add the package back into the repo compressed, but it's still too big to include enabled by default obviously.  if you compile your own build, which is very easy, you'll be able to include it.

**Reply by gtxaspec** [reply]
another issue is i can't get tailscale to compile from source using the buildroot system...

**Reply by gtxaspec** [reply]
ok i figured it out. compressed, the combined binary tailscale is now 4.4mb.
thanks to @numericOverflow for the link to small tailscale.

**Comment by gtxaspec** [comment]
added back in:
https://github.com/themactep/thingino-firmware/commit/cec0c1db0f5581b8573da6a4460dcde467975b8e
enable with `BR2_PACKAGE_THINGINO_TAILSCALE=y`

**Reply by numericOverflow** [reply]
Cool.  

I'm not familiar with how this is ultimately included in the firmware.  I was going to try it out but I'm not seeing it on my cam. Do I need to compile my own version, or what's needed to get the binary on the cam this way?

**Reply by gtxaspec** [reply]
yes, you need to compile your own version with that option enabled.

**Comment by aou** [comment]
I got this working with a patch to the package defs:

```
-define TAILSCALE_LINUX_CONFIG_FIXUPS
-	$(call KCONFIG_ENABLE_OPT,CONFIG_TUN)
+define THINGINO_TAILSCALE_LINUX_CONFIG_FIXUPS
+	$(call KCONFIG_SET_OPT,CONFIG_TUN,m)
```

Think there was a typo.

However, tailscaled keeps getting killed due to OOM. I worked around this with a custom launcher that would respawn on non-zero exit, but seems pretty spotty, so I went back to vanilla wireguard.

Is there a way to reduce memory usage? I'm running pretty much stock firmware + tailscale on a wyze cam v3.

**Comment by andypiper** [comment]
I am thoroughly confused.

No matter where I add `BR2_PACKAGE_THINGINO_TAILSCALE` this does not seem to add it.

I'd also like to be able to disable Telegram, since I never use that and it can save some memory.

**Reply by sstepansky** [reply]
Support was dropped 
https://github.com/themactep/thingino-firmware/commit/e7ac24eb3006bedcb4949b1ec9ec2afbc1e3eedc

**Reply by fencepost** [reply]
I think this is an area actively changing based on https://github.com/themactep/thingino-firmware/commit/31200afed1b75c9d26d28b9318857ef8f45dc2ab

**Reply by gtxaspec** [reply]
Tailscale is now included in buildroot natively, so this is easier now.  Live in master, needs testing.

**Reply by andypiper** [reply]
happy to test, failing to find! Should be in `make menuconfig` under Extra Packages AFAICT, but not showing up for my Cinnado :-/

**Comment by rrrainielll** [comment]
Update: I attempted to install Tailscale on the Xiaomi MJSXJ03HL using the small binary, but it fails due to insufficient RAM, resulting in an OOM (Out of Memory) error. The failure appears to occur during decompression, as even the small binary requires more memory than the device can provide.

I also tried placing the small binary on the SD card, but the same OOM error occurs. Next, I tested running the already uncompressed binary directly from the SD card, which also led to a dead end.

Additionally, I later confirmed that the firmware does not support TUN/TAP. Without TUN support, there is no viable way to integrate or run Tailscale on this device at the firmware level.

**Reply by gtxaspec** [reply]
The firmware needs to be compiled with the tailscale binary, this should enable tun/tap support (check the kernel configs in the repo)
I would also not recommend running it on devices with only 64mb of ram, 128mb only.
